introduction-jsx translation (#10)
reactjs/fr.reactjs.org: content/docs/introducing-jsx.md

---
id: introducing-jsx
title: Introduction à JSX
permalink: docs/introducing-jsx.html
prev: hello-world.html
next: rendering-elements.html
---

> Observez cette déclaration de variable :

Author: Tu voudrais dire « Prenons cette déclaration de variable » ? Ou « Prenons cette déclaration de variable comme exemple »

> Observez cette déclaration de variable :

Reviewer: La première 😁

```js
const element = <h1>Bonjour, monde !</h1>;
```

Cette drôle de syntaxe n'est ni une chaîne de caractères ni du HTML. 
 

> Cela s'appelle du JSX, c'est une extension syntaxique de JavaScript. Nous recommandons de l'utiliser avec React afin de décrire à quoi devrait ressembler l'interface utilisateur. JSX vous fait surement penser à un langage de balisage, mais il est accompagné de tout le pouvoir de JavaScript.

Author: - Désolé pour les "cela", j'avais pourtant lu les consignes mais c'est trop un reflex. 
- D'accord pour UI, on devrait ajouter cette note au glossaire ?
- Concernant "pouvoir" j'avais hésité avec "la force de" mais ta nouvelle tournure est meilleure en effet.

> JSX produit des "éléments" React. Nous explorerons comment les rendre dans le DOM dans la [prochaine section](/docs/rendering-elements.html). Ci-dessous vous pouvez trouver les bases en JSX à savoir afin de bien démarrer.

Author: J’étais perturbé par le mot "render", j'avoue m’être basé sur l'infinitif de la traduction de "render props" mais c'est pas du tout le même "render", merci !

> ### Pourquoi le JSX ? {#why-jsx}

Author: Ok je comprends, navré j'ai fait cette erreur plusieurs fois.

> React embrasse le fait que la logique de rendu est intrinsèquement couplée à la logique de l'interface utilisateur : comment les événements sont-ils gérés, comment l'état évolue t-il dans le temps, et comment la donnée est-elle préparée pour l'affichage.

Author: C'est correct, merci pour tes feedback et ta patience. J'avais peur de trop m’éloigner du texte de base en réécrivant mais au final c'est devenue une traduction mot-à-mot.

> Au lieu de séparer artificiellement les *technologies* en mettant le balisage et la logique dans des fichiers séparés, React [sépare les *préoccupations*](https://fr.wikipedia.org/wiki/S%C3%A9paration_des_pr%C3%A9occupations) via des unités faiblement couplées appelées des "composants" qui contiennent les deux. Nous reviendrons sur les composants dans une [prochaine section](/docs/components-and-props.html), si vous vous ne sentez pas convaincu avec l'idée de mixer des balises dans du JS, [cette présentation](https://www.youtube.com/watch?v=x7cQ3mrcKaY) pourra peut-être vous faire changer d'avis.

Reviewer: ```suggestion
Au lieu de séparer artificiellement les *technologies* en mettant le balisage et la logique dans des fichiers séparés, React [sépare les *préoccupations*](https://fr.wikipedia.org/wiki/S%C3%A9paration_des_pr%C3%A9occupations) via des unités faiblement couplées appelées « composants », qui contiennent les deux. Nous reviendrons sur les composants dans une [prochaine section](/docs/components-and-props.html), mais si l'idée d’injecter des balises dans du JS vous met mal à l’aise, [cette présentation](https://www.youtube.com/watch?v=x7cQ3mrcKaY) vous fera peut-être changer d'avis.
```

OK, alors là tu vois y'a très peu à retoucher, tu as fait du bon boulot ! 😎 

- [Guillemets français](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#guillemets-français)
- Fluidification / allègement de certaines formules

> React ne [nécessite pas](/docs/react-without-jsx.html) l'utilisation de JSX pour fonctionner, mais la plupart des gens le considère comme une aide visuel quand ils manipulent l'interface utilisateur dans le code JavaScript. Cela permet aussi à React de montrer des messages d'erreurs et d'avertissements plus utiles.

Reviewer: ```suggestion
React [ne vous oblige pas](/docs/react-without-jsx.html) à utiliser JSX, mais la plupart des gens y trouvent une aide visuelle quand ils manipulent l'interface utilisateur dans le code JavaScript. Ça permet aussi à React de produire des messages d'erreurs et d'avertissements plus utiles.
```

Pas mal !

- La première phrase est avant tout une question de contrainte imposée ou non, on a donc reformulé dans ce sens. - De façon générale, on allège un peu.
- Plus spécifiquement, « Cela » est très formel et rarement pertinent, on préfère généralement « ça ».
- « Montrer » est exact en soi, mais n'est pas très élégant de façon générale, d'autant que si ta console est masquée, React ne « montre » rien. C'est pourquoi j'ai aussi évité le plus classique « afficher », pour aller vers un plus générique « produire ».

> Nous pouvons enfin commencer !

Reviewer: ```suggestion
Ceci étant posé, commençons !
```

Ton « enfin » traduisait l'idée qu'on n'a pas pu démarrer tout de suite, et l'excitation d'y aller ; l'intention est bonne. En revanche, il traduit aussi une impatience, une exaspération, le tout sans justifier le délai au démarrage, justif' qu'on trouvait dans le “With that out of the way”, qui sous-entend un détour qui était nécessaire.

> ### Imbriqué des Expressions en JSX {#embedding-expressions-in-jsx}

Reviewer: ```suggestion
### Utiliser des expressions dans JSX {#embedding-expressions-in-jsx}
```

Outre la faute d'accord, « imbriquer » est souvent maladroit quand on parle de code. Et puis il sous-entend que le contenant et le contenu sont de nature similaire. Ici, JSX est un contexte au sein duquel on veut pouvoir utiliser les expressions, autant utiliser un verbe simple.

*(Pour la casse, la règle de capitalisation vient de changer pour favoriser la capitalisation naturelle de la langue cible de la traduction)*

> Dans l'exemple suivant, nous déclarons une variable appelée `name` et nous l'utilisons ensuite dans du JSX en l'encadrant avec des accolades :

Reviewer: ```suggestion
Dans l'exemple suivant, nous déclarons une variable appelée `name` et nous l'utilisons ensuite dans JSX en l'encadrant avec des accolades :
```

Impec’, à ceci près qu'on va toujours considérer JSX comme un contexte, pas comme une matière, un sujet, etc. Dans la plupart du temps, pas d'article nécessaire. Ou alors, « le » ici, puisqu'on parle d'un bloc JSX précis.

```js{1,2}

> const name = 'Clarisse Agbegnenou';

Reviewer: 💯 pour le choix de trad d'identité 😁

Author: Ça reste centré sur la France cependant, je pense pas que mes collègues Montréalais la connaissent.

Reviewer: Héhé, en même temps les Montréalais ont infiniment moins de mal à lire la VO que le Français moyen 😞

const element = <h1>Bonjour, {name}</h1>;

ReactDOM.render(
  element,
  document.getElementById('root')
);
```

> Vous pouvez utiliser n'importe quelle [expression JavaScript](https://developer.mozilla.org/fr/docs/Web/JavaScript/Guide/Expressions_et_Op%C3%A9rateurs#Expressions) valide dans des accolades en JSX. Par exemple, `2 + 2`, `user.firstName`, or `formatName(user)` sont toutes des expressions JavaScript valides.

Reviewer: ```suggestion
Vous pouvez utiliser n'importe quelle [expression JavaScript](https://developer.mozilla.org/fr/docs/Web/JavaScript/Guide/Expressions_et_Op%C3%A9rateurs#Expressions) valide dans des accolades en JSX. Par exemple, `2 + 2`, `user.firstName`, ou `formatName(user)` sont toutes des expressions JavaScript valides.
```

Le *Syndrôme de la Conjonction Oubliée Dans le Feu de l’Action™*, bien connu… 😉

> [](codepen://introducing-jsx)

Reviewer: ```suggestion
**[Essayer sur CodePen](codepen://introducing-jsx)**
```

Le texte par défaut du *CodePen REPL* sera en anglais et en gras, la bonne manière de recaler la trad en préservant la graisse [est celle-ci](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#liens-codepen).

> On sépare le JSX sur plusieurs lignes pour une meilleure lecture. Par la même occasion, nous recommandons aussi de le mettre entre parenthèses afin d'éviter les pièges [d'insertion de point-virgule automatique](http://stackoverflow.com/q/2846283), notez que cette pratique n'est pas obligatoire.

Reviewer: ```suggestion
On découple le JSX en plusieurs lignes pour une meilleure lisibilité. Par la même occasion, nous recommandons également de le mettre entre parenthèses afin d'éviter les pièges [d'insertion de point-virgule automatique](http://stackoverflow.com/q/2846283), même si cette pratique n'est pas obligatoire.
```

Presque impeccable 😁

> ### JSX est aussi une expression {#jsx-is-an-expression-too}

Reviewer: ```suggestion
### JSX n’est rien d’autre qu’une expression {#jsx-is-an-expression-too}
```

La formule usuelle anglaise n'existe pas en mot-à-mot en français, du coup on revient sur l'intention.

> Après la compilation, les expressions JSX deviennent des appels de fonctions JavaScripts réguliers et sont évalués en tant qu'objet JavaScript.

Reviewer: ```suggestion
Après la compilation, les expressions JSX deviennent de simples appels de fonctions JavaScripts, dont l'évaluation renvoie des objets JavaScript.
```

- Petit glissement sémantique, les fonctions « ne sont pas évaluées en tant que », mais leur évaluation produit comme résultat.
- Attention à “regular”, qui 99% du temps veut dire « normal », « habituel », « classique », et pas « régulier » (même pour “regular expression”, dont la VF dominante est « expression rationnelle »). Par ailleurs, le français préfère les adjectifs en position épithète ou attribut.

> Cela signifie que vous pouvez utiliser JSX à l'intérieur d'instructions `if` ou de boucles `for`, l'assigner dans une variable, l'accepter en tant qu'argument, et le retourner depuis des fonctions :

Reviewer: ```suggestion
Ça signifie que vous pouvez utiliser JSX à l'intérieur d'instructions `if` ou de boucles `for`, l'affecter à des variables, l'accepter en tant qu'argument, et le renvoyer depuis des fonctions :
```

- Toujours le Cela -> Ça, bien sûr
- [Affecter, pas Assigner](https://github.com/reactjs/fr.reactjs.org/wiki/Glossaire)
- Une fonction « renvoie » en français, elle ne « retourne » pas 😉

```js{3,5}
function getGreeting(user) {
  if (user) {
    return <h1>Bonjour, {formatName(user)} !</h1>;
  }

>   return <h1>Bonjour, Inconnu.</h1>;

Reviewer: ```suggestion
  return <h1>Bonjour, Belle Inconnue.</h1>;
```

Dans la culture populaire anglaise / US, “Hello, Stranger” est généralement prononcée de façon ironique et intime en face d'un être aimé, le plus souvent absent·e depuis longtemps.

}
```

> ### Spécifier des Attributs avec JSX {#specifying-attributes-with-jsx}

Reviewer: ```suggestion
### Spécifier des attributs en JSX {#specifying-attributes-with-jsx}
```

> Vous pouvez utiliser des guillemets doubles pour spécifier des chaînes de caractères littérales en tant qu’attributs :

Reviewer: ```suggestion
Vous pouvez utiliser des guillemets pour spécifier des littéraux chaînes de caractères dans les attributs :
```

Ce n'est pas la chaîne qui est littérale (ce qui sous-entend que son *contenu* est à interpréter de façon *littérale*, sur le fond), c'est un littéral *l’élément syntaxique de langage*, terme toujours masculin, qu'on qualifie du coup avec son type « chaîne de caractères » (je déteste la longueur de ce truc, c'est pas pour rien qu'à l'oral on dit juste « chaîne »…).

Par ailleurs, JSX autorise tant les guillemets simples que doubles, on n'est pas dans cette !@# de JSON 😉 

Le “as attributes” est un raccourci malheureux parce qu'on parle des valeurs d'attributs, il faut aussi leur nom.

```js
const element = <div tabIndex="0"></div>;
```

> Vous pouvez aussi utiliser des accolades pour intégrer une expression JavaScript dans un attribut :

Reviewer: ```suggestion
Vous pouvez aussi utiliser des accolades pour utiliser une expression JavaScript dans un attribut :
```

```js
const element = <img src={user.avatarUrl}></img>;
```

> Ne mettez pas des guillemets doubles autour d'accolades quand vous intégrez une expression JavaScript dans un attribut. You devrez utiliser des guillemets doubles (for des valeurs de type chaîne de caractères) ou des accolades (pour des expressions), mais pas les deux dans le même attribut.

Reviewer: ```suggestion
Ne mettez pas de guillemets autour des accolades quand vous utilisez une expression JavaScript dans un attribut. Vous pouvez utiliser soit des guillemets (pour des valeurs textuelles) soit des accolades (pour des expressions), mais pas les deux à la fois pour un même attribut.
```

Marrants ces petits oublis de trad en cours de route 😁
Pour le reste, fluifidication. Notamment, on traduit généralement *should* soit de façon forte (« devez », le *should* est poli alors que c'est en vrai “must”), soit de façon douce (« pouvez », le *should* étant une reco non obligatoire, le contexte de la doc officielle rend la reco implicite, et le « pourriez » ou « devriez » rend vraiment trop moche).

Author: C'est clairement un manque de relecture dans la précipitation, désolé !

>**Attention :**
>

> >Comme JSX est plus proche de JavaScript qu'il ne l'est du HTML, React DOM utilise le `camelCase` comme convention de nommage pour ses propriétés au lieu des noms des attributs HTML.

Reviewer: ```suggestion
>Dans la mesure où JSX est plus proche de JavaScript que de HTML, React DOM utilise la casse `camelCase` comme convention de nommage des propriétés, au lieu des noms d’attributs HTML.
```

« Comme » en début de phrase est vraiment moche en français. Et allègement.

>
>Par exemple, `class` devient [`className`](https://developer.mozilla.org/fr/docs/Web/API/Element/className) en JSX, et `tabindex` devient [`tabIndex`](https://developer.mozilla.org/fr/docs/Web/API/HTMLElement/tabIndex).

> ### Spécifier des Enfants avec JSX {#specifying-children-with-jsx}

Reviewer: ```suggestion
### Spécifier des éléments enfants en JSX {#specifying-children-with-jsx}
```

Ça mérite d'être plus explicite (on pourrait le proposer *upstream* aussi)

Author: Je pensais aussi au mot descendant.

Reviewer: Sauf que ça n'a pas du tout le même sens. Les enfants sont le premier niveau, au sens React en tout cas (ceux qui sont spécifiés dans le `render()` ou la fonction composant), les *descendants* sont à n'importe quel niveau de profondeur en-dessous.

> Si une balise est vide, vous pouvez le fermer immédiatement via `/>`, comment en XML :

Reviewer: ```suggestion
Si une balise est vide, vous pouvez la fermer immédiatement avec `/>`, comme en XML :
```

```js
const element = <img src={user.avatarUrl} />;
```

Les balises JSX peuvent contenir des enfants :

```js
const element = (
  <div>
    <h1>Bonjour !</h1>
    <h2>Content de te voir ici.</h2>
  </div>
);
```

> ### JSX Empêchent les Attaques de type Injection {#jsx-prevents-injection-attacks}

Reviewer: ```suggestion
### JSX empêche les attaques d’injection {#jsx-prevents-injection-attacks}
```

> Vous ne risquez rien si vous intégré une entrée utilisateur en JSX :

Reviewer: ```suggestion
Vous ne risquez rien en utilisant une saisie utilisateur dans JSX :
```

```js
const title = response.potentiallyMaliciousInput;

> // Cela est sûr :

Reviewer: ```suggestion
// Ceci est sans risque :
```

const element = <h1>{title}</h1>;
```

> Par défaut, React DOM [échappe](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html) toutes les valeurs intégrées dans JSX avant de les rendre. Cela assure que vous ne pouvez jamais injecter quoi que ce soit d'autre que ce vous avez explicitement écrit dans votre application. Tout est converti en chaîne de caractères avant d'être rendu. Celle contribute a la prevention des attaques de type [XSS (cross-site-scripting)](https://fr.wikipedia.org/wiki/Cross-site_scripting).

Reviewer: ```suggestion
Par défaut, React DOM [échappe](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html) toutes les valeurs intégrées avec JSX avant d’en faire le rendu. Il garantit ainsi que vous ne risquez jamais d’injecter quoi que ce soit d'autre que ce vous avez explicitement écrit dans votre application. Tout est converti en chaîne de caractères avant de produire le rendu. Ça aide à éviter les attaques [XSS (cross-site-scripting)](https://fr.wikipedia.org/wiki/Cross-site_scripting).
```

- On [évite autant que possible](https://github.com/reactjs/fr.reactjs.org/wiki/Glossaire#render-re--renderer) « rendre » comme verbe
- Petits foirages de trad'

> ### JSX Représente des Objets {#jsx-represents-objects}

Reviewer: ```suggestion
### JSX représente des objets {#jsx-represents-objects}
```

> Babel compile JSX vers des appels `React.createElement()`.

Reviewer: ```suggestion
Babel compile JSX vers des appels à `React.createElement()`.
```

const element = React.createElement(
  'h1',
  {className: 'greeting'},
  'Bonjour, monde !'
);
```

> `React.createElement()` effectue quelques controls pour vous aider à écrire un code sans bug mais essentiellement il crée un semblant d'objet qui ressemble à ceci :

Reviewer: ```suggestion
`React.createElement()` effectue quelques vérifications pour vous aider à écrire un code sans bug, mais pour l'essentiel il crée un objet qui ressemble à ceci :
```

```js
// Note: this structure is simplified
const element = {
  type: 'h1',
  props: {
    className: 'greeting',
    children: 'Bonjour, monde !'
  }
};
```

> Ces objets sont appelés des "éléments React". Vous pouvez les imaginer commes des descriptions de ce que vous voulez voir sur l'écran. React lit ces objets et les utilises pour construir le DOM et tenir à jour.

Reviewer: ```suggestion
Ces objets sont appelés des « éléments React ». Vous pouvez les considérer comme des descriptions de ce que vous voulez voir sur l'écran. React lit ces objets et les utilise pour construire le DOM et le tenir à jour.
```

- Guillemets français
- Quelques couacs de trad'

Author: Désolé j;ai mal relu

> Nous explorons le rendu des éléments React vers le DOM dans la prochaine section.

Reviewer: ```suggestion
Nous explorerons la retranscription des éléments React dans le DOM dans la prochaine section.
```

En temps normal on éviterait « explorer » pour lui préférer « voir », mais là ça passe, juste, concordance des temps. En revanche, ici « rendu » est trop proche d'une utilisation en verbe plutôt qu'en nom.

>**Astuce :**
>

> >Nous recommandons d'utiliser le [language de définition "Babel"](http://babeljs.io/docs/editors) dans votre éditeur préféré ainsi les codes ES6 et JSX sont proprement soulignés. Ce site internet utilise schéma de couleur [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/) qui est compatible avec celui-ci.

Reviewer: ```suggestion
>Nous recommandons d'utiliser la [définition de langage « Babel »](http://babeljs.io/docs/editors) dans votre éditeur préféré, afin que les codes ES6 et JSX soient correctement colorisés. Ce site utilise le thème de couleurs [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/), qui est compatible avec ce mode syntaxique.
```

Hmmm, quelques plantages sémantiques. Mais on a fini ! 🎉

Author: 🎉 merci encore
